include/cuco/detail/cache_residency_control.cuh

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ */
+
+#pragma once
+
+#include <algorithm>
+#include <iterator>
+
+namespace cuco {
+
+/**
+ * @brief Registers the global memory region `[begin, end)` to
+ * be permanently resident in L2 cache.
+ *
+ * @tparam Iterator Accessor of the memory region
+ * @param[in, out] stream The CUDA stream this region is accessed through
+ * @param[in] begin Start of the memory region to be mapped
+ * @param[in] end End of the memory region
+ * @param[in] hit_rate Probability for a sub-segment to be mapped in L2
+ * @param[in] carve_out Fraction of total L2 space to be blocked for resident memory segments
+ *
+ * @note Only has effect on Ampere and above.
+ * @note Assumes the memory region to be contiguous.
+ */
+template <typename Iterator>
+void register_l2_persistence(
+  cudaStream_t& stream, Iterator begin, Iterator end, float hit_rate = 0.6f, float carve_out = 1.0f)
```

Comment:
For now we use an iterator pair and assume they represent a continuous memory segment. This is going to look a lot cleaner once we can use span.

Reply:
Can you include a description of why this is necessary vs. `annotated_ptr`?

Reply:
The most elegant solution to this problem would be to wrap the storage of the filter as an `annotated_ptr`. However, this first requires a solution to the `annotated_ptr` + `cuda::atomic`/`cuda::atomic_ref` problem, which is not available yet.

The next best approach would be to set the access property manually using `cuda::apply_access_property`, which I implemented in the previous version of the benchmark script: https://github.com/NVIDIA/cuCollections/blob/a1ea293ff4ce4c3bbbcd1aaecb415b75dec4c5bb/benchmarks/bloom_filter/bloom_filter_bench.cu#L199-L217

This however showed little to no performance impact. I have not been able to quite get to the bottom of why this is happening. Another disadvantage of this method is that `cuda::apply_access_property` does not allow hybrid access policies, e.g. 60% `cuda::access_property_persisting` + 40% `cuda::access_property_normal`, which showed the best performance in the benchmarks.

tl;dr This is a workaround that is flexible enough for our purposes (can also be used with other cuco map types) but should be replaced once `annotated_ptr` works as expected.

Reply:
> Could you place `stream` as the last parameter? Not necessary but try to be consistent with other cuco APIs.

Yes, I can do that for the sake of consistency. I initially put it in the front since it's an in/out parameter, i.e., the stream attributes are mutated.

However, this reminds me that I find the placement of the stream parameter in our API unfortunate. I would put the stream in front of the hashers and equality operators, since you are likely to set this one explicitly more often than the others.
It also has a shorter default value compared to the others.

Reply:
Wait, I can't put the stream at the since there are additional defaulted parameters.

Reply:
> The most elegant solution to this problem would be to wrap the storage of the filter as an `annotated_ptr`. However, this first requires a solution to the `annotated_ptr` + `cuda::atomic`/`cuda::atomic_ref` problem, which is not available yet.
> 
> The next best approach would be to set the access property manually using `cuda::apply_access_property`, which I implemented in the previous version of the benchmark script:
> 
> https://github.com/NVIDIA/cuCollections/blob/a1ea293ff4ce4c3bbbcd1aaecb415b75dec4c5bb/benchmarks/bloom_filter/bloom_filter_bench.cu#L199-L217
> 
> This however showed little to no performance impact. I have not been able to quite get to the bottom of why this is happening. Another disadvantage of this method is that `cuda::apply_access_property` does not allow hybrid access policies, e.g. 60% `cuda::access_property_persisting` + 40% `cuda::access_property_normal`, which showed the best performance in the benchmarks.
> 
> tl;dr This is a workaround that is flexible enough for our purposes (can also be used with other cuco map types) but should be replaced once `annotated_ptr` works as expected.

Hey, apologies for bringing up this old PR. I am also testing the cuda::apply_access_property API and in general the prefetch.global.L2::evict_last PTX instruction that it internally uses, but I am not able to see any effects of this. However,  when I use the accessPolicyWindow method from the CUDA API to configure L2 persistence for the same scenario, I do observe performance improvements. Did you figure out the reason or any possible resolution for this issue? Any help would be greatly appreciated.